flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,614 @@
+ * of the algorithm and the [[FitOperation]] & [[PredictOperation]].
+ */
+object KMeans {
+
+  private val RANDOM_FRACTION = "random_sample_fraction"
+  private val PARINIT_SET = "par_init_solution_set"
+  private val PARINIT_COST = "par_init_solution_cost"
+  private val PARINIT_SAMPLE = "par_init_oversample_factor"
+
+  /** Euclidean Distance Metric */
+  val euclidean = EuclideanDistanceMetric()
+
+  case object NumIterations extends Parameter[Int] {
+    val defaultValue = Some(10)
+  }
+
+  case object InitialCentroids extends Parameter[Seq[LabeledVector]] {
+    val defaultValue = None
+  }
+
+  case object InitialStrategy extends Parameter[String]{
+    val defaultValue = Some("kmeans||")
+  }
+
+  case object NumClusters extends Parameter[Int] {
+    val defaultValue = None
+  }
+
+  case object OversamplingFactor extends Parameter[Double] {
+    val defaultValue = None
+  }
+
+  case object KMeansParRounds extends Parameter[Int] {
+    val defaultValue = Some(5)
+  }
+
+  // ========================================== Factory methods ====================================
+
+  def apply(): KMeans = {
+    new KMeans()
+  }
+
+  // ========================================== Operations =========================================
+
+  /** Provides the operation that makes the predictions for individual examples.
+    * The label of the vector will be the index of the cluster the input vector belongs to.
+    *
+    * @tparam T
+    * @return A PredictOperation, through which it is possible to predict a value, given a
+    *         feature vector
+    */
+  implicit def predictVectors[T <: Vector] = {
+    new PredictOperation[KMeans, Seq[LabeledVector], T, Double](){
+
+      override def getModel(
+          self: KMeans,
+          predictParameters: ParameterMap)
+        : DataSet[Seq[LabeledVector]] = {
+
+        self.centroids match {
+          case Some(model) => model
+          case None => {
+            throw new RuntimeException("The KMeans model has not been trained. Call first fit" +
+              "before calling the predict operation.")
+          }
+        }
+      }
+
+      override def predict(value: T, model: Seq[LabeledVector]): Double = {
+        findNearestCentroid(value, model)._1
+      }
+    }
+  }
+
+  /**
+   * [[FitOperation]] which iteratively computes centroids that match the given input DataSet by
+   * adjusting the given initial centroids.
+   *
+   * @return A new  [[FitOperation]] to train the model using the training data set.
+   */
+  implicit def fitKMeans = {
+    new FitOperation[KMeans, Vector] {
+      override def fit(instance: KMeans, fitParameters: ParameterMap, trainingDS: DataSet[Vector])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+
+        // =================  INITIALIZATION OF KMEANS ==========================
+        val centroids: DataSet[Seq[LabeledVector]] = init(trainingDS, resultingParameters)
```

Comment:
Why do we store the centroids in a `DataSet` with only a single element of type `Seq[LabeledVector]`. Why not distributedly in a `DataSet[LabeledVector]`?